Multi-Turn Rollout on a single GPU

## TLDR;
Running into the assertion-error at `/verl/workers/rollout/vllm_rollout/vllm_async_server.py:55`
when running a single node variant of `verl/tests/rollout/test_vllm_multi_turn.py`. (See script in Appendix A)

## Description 

Given the exciting work on multi-turn rollouts introduced in https://github.com/volcengine/verl/pull/1138, we are trying to add multimodal environment interactions (similar idea as described in the [VAGEN](https://mll-lab.notion.site/vagen)-paper)). 

However, I run into an assertion error when running the `AsyncLLMManager` with a single `AsyncLLMWorker`. I tried going through the debugger for quite some while but cannot figure out why the Executor does not seem to find the Actors.

##  Reproduction


(see Appendix A for script)
```terminal
python tests/rollout/test_vllm_multi_turn.py
```

```bash
2025-05-02 14:57:34,894 DEBUG worker.py:1576 -- Automatically increasing RLIMIT_NOFILE to max value of 65536
2025-05-02 14:57:34,897 DEBUG node.py:293 -- Setting node ID to d1cce7a158262a4f3567d99a4612737ffef92c57256a3bc1d555541b
2025-05-02 14:57:34,903 DEBUG node.py:1401 -- Process STDOUT and STDERR is being redirected to /tmp/ray/session_2025-05-02_14-57-34_895398_2793/logs.
2025-05-02 14:57:35,829 DEBUG node.py:1430 -- Process STDOUT and STDERR is being redirected to /tmp/ray/session_2025-05-02_14-57-34_895398_2793/logs.
2025-05-02 14:57:35,861 DEBUG npu.py:60 -- Could not import AscendCL: No module named 'acl'
2025-05-02 14:57:35,863 DEBUG tpu.py:115 -- Failed to detect number of TPUs: [Errno 2] No such file or directory: '/dev/vfio'
2025-05-02 14:57:35,864 DEBUG services.py:2140 -- Determine to start the Plasma object store with 9.49 GB memory using /dev/shm.
2025-05-02 14:57:35,912 INFO worker.py:1832 -- Started a local Ray instance. View the dashboard at 127.0.0.1:8265 
2025-05-02 14:57:36,852 WARNING __init__.py:161 -- DeprecationWarning: `ray.state.available_resources_per_node` is a private attribute and access will be removed in a future Ray version.
colocated worker base class <class 'verl.single_controller.base.worker.Worker'>
bind role actor_rollout method execute_method to class <class 'verl.single_controller.ray.base.create_colocated_worker_cls.<locals>.WorkerDict'>
WARNING:2025-05-02 14:57:37,950:Waiting for register center actor 5Y8l0k_register_center to be ready. Elapsed time: 0 seconds out of 300 seconds.
(WorkerDict pid=3433) Model config after override: Qwen2Config {
(WorkerDict pid=3433)   "architectures": [
(WorkerDict pid=3433)     "Qwen2ForCausalLM"
(WorkerDict pid=3433)   ],
(WorkerDict pid=3433)   "attention_dropout": 0.0,
(WorkerDict pid=3433)   "eos_token_id": 151645,
(WorkerDict pid=3433)   "hidden_act": "silu",
(WorkerDict pid=3433)   "hidden_size": 896,
(WorkerDict pid=3433)   "initializer_range": 0.02,
(WorkerDict pid=3433)   "intermediate_size": 4864,
(WorkerDict pid=3433)   "max_position_embeddings": 32768,
(WorkerDict pid=3433)   "max_window_layers": 21,
(WorkerDict pid=3433)   "model_type": "qwen2",
(WorkerDict pid=3433)   "num_attention_heads": 14,
(WorkerDict pid=3433)   "num_hidden_layers": 24,
(WorkerDict pid=3433)   "num_key_value_heads": 2,
(WorkerDict pid=3433)   "pad_token_id": 151643,
(WorkerDict pid=3433)   "rms_norm_eps": 1e-06,
(WorkerDict pid=3433)   "rope_scaling": null,
(WorkerDict pid=3433)   "rope_theta": 1000000.0,
(WorkerDict pid=3433)   "sliding_window": 32768,
(WorkerDict pid=3433)   "tie_word_embeddings": true,
(WorkerDict pid=3433)   "torch_dtype": "bfloat16",
(WorkerDict pid=3433)   "transformers_version": "4.51.1",
(WorkerDict pid=3433)   "use_cache": true,
(WorkerDict pid=3433)   "use_sliding_window": false,
(WorkerDict pid=3433)   "vocab_size": 151936
(WorkerDict pid=3433) }
(WorkerDict pid=3433) 
(WorkerDict pid=3433) Flash Attention 2.0 only supports torch.float16 and torch.bfloat16 dtypes, but the current dype in Qwen2ForCausalLM is torch.float32. You should run training or inference using Automatic Mixed-Precision via the `with torch.autocast(device_type='torch_device'):` decorator, or load the model with the `torch_dtype` argument. Example: `model = AutoModel.from_pretrained("openai/whisper-tiny", attn_implementation="flash_attention_2", torch_dtype=torch.float16)`
(WorkerDict pid=3433) You are attempting to use Flash Attention 2.0 with a model not initialized on GPU. Make sure to move the model to GPU after initializing it on CPU with `model.to('cuda')`.
(WorkerDict pid=3433) NCCL version 2.21.5+cuda12.4
(WorkerDict pid=3433) [rank0]:[W502 14:57:46.146030738 ProcessGroupNCCL.cpp:4561] [PG ID 0 PG GUID 0 Rank 0]  using GPU 0 to perform barrier as devices used by this process are currently unknown. This can potentially cause a hang if this rank to GPU mapping is incorrect. Specify device_ids in barrier() to force use of a particular device, or call init_process_group() with a device_id.
(WorkerDict pid=3433) Qwen2ForCausalLM contains 494.03M parameters
(WorkerDict pid=3433) wrap_policy: functools.partial(<function _or_policy at 0x7f8717a2e170>, policies=[functools.partial(<function transformer_auto_wrap_policy at 0x7f8717a2e050>, transformer_layer_cls={<class 'transformers.models.qwen2.modeling_qwen2.Qwen2DecoderLayer'>})])
(WorkerDict pid=3433) /usr/local/lib/python3.10/dist-packages/torch/distributed/fsdp/_init_utils.py:444: UserWarning: FSDP is switching to use `NO_SHARD` instead of ShardingStrategy.FULL_SHARD since the world size is 1.
(WorkerDict pid=3433)   warnings.warn(
(WorkerDict pid=3433) Total steps: -1, num_warmup_steps: 0
(WorkerDict pid=3433) Actor use_remove_padding=False
(WorkerDict pid=3433) /usr/local/lib/python3.10/dist-packages/torch/distributed/fsdp/fully_sharded_data_parallel.py:690: FutureWarning: FSDP.state_dict_type() and FSDP.set_state_dict_type() are being deprecated. Please use APIs, get_state_dict() and set_state_dict(), which can support different parallelisms, FSDP1, FSDP2, DDP. API doc: https://pytorch.org/docs/stable/distributed.checkpoint.html#torch.distributed.checkpoint.state_dict.get_state_dict .Tutorial: https://pytorch.org/tutorials/recipes/distributed_checkpoint_recipe.html .
(WorkerDict pid=3433)   warnings.warn(
INFO 05-02 14:57:49 [__init__.py:239] Automatically detected platform cuda.
(AsyncvLLMServer pid=3620) FastAPI startup
(AsyncvLLMServer pid=3620) override_generation_config: {'n': 1, 'logprobs': 0, 'max_tokens': 1024, 'temperature': 1.0, 'top_k': -1, 'top_p': 1, 'ignore_eos': False}
(AsyncvLLMServer pid=3620) WARNING 05-02 14:58:04 [arg_utils.py:1713] Detected VLLM_USE_V1=1 with Engine in background thread. Usage should be considered experimental. Please report any issues on Github.
(AsyncvLLMServer pid=3620) WARNING 05-02 14:58:04 [cuda.py:96] To see benefits of async output processing, enable CUDA graph. Since, enforce-eager is enabled, async output processor cannot be used
(AsyncvLLMServer pid=3620) WARNING 05-02 14:58:04 [core_client.py:368] SIGUSR1 handler not installed because we are not running in the main thread. In this case the forked engine process may not be killed when an exception is raised, and you need to handle the engine process shutdown manually.
(AsyncvLLMServer pid=3620) 2025-05-02 14:58:11,322      INFO worker.py:1832 -- Started a local Ray instance. View the dashboard at 127.0.0.1:8266 
(AsyncvLLMServer pid=3620) EngineCoreProc.run_engine_core
(AsyncvLLMServer pid=3620) EngineCoreProc.__init__
(AsyncvLLMServer pid=3620) EngineCore.__init__ - Start
(AsyncvLLMServer pid=3620) EngineCore.__init__ - Before model executor initialization
(AsyncvLLMServer pid=3620) namespace: 39fff769-940c-4c88-bb08-0ad63fb5e321
(AsyncvLLMServer pid=3620) actor_names: []
(AsyncvLLMServer pid=3620) all actors: []
(AsyncvLLMServer pid=3620) ERROR 05-02 14:58:12 [core.py:406] EngineCore hit an exception: Traceback (most recent call last):
(AsyncvLLMServer pid=3620) ERROR 05-02 14:58:12 [core.py:406]   File "/home/ec2-user/vllm/vllm/v1/engine/core.py", line 394, in run_engine_core
(AsyncvLLMServer pid=3620) ERROR 05-02 14:58:12 [core.py:406]     engine_core = EngineCoreProc(*args, **kwargs)
(AsyncvLLMServer pid=3620) ERROR 05-02 14:58:12 [core.py:406]   File "/home/ec2-user/vllm/vllm/v1/engine/core.py", line 335, in __init__
(AsyncvLLMServer pid=3620) ERROR 05-02 14:58:12 [core.py:406]     super().__init__(vllm_config, executor_class, log_stats)
(AsyncvLLMServer pid=3620) ERROR 05-02 14:58:12 [core.py:406]   File "/home/ec2-user/vllm/vllm/v1/engine/core.py", line 68, in __init__
(AsyncvLLMServer pid=3620) ERROR 05-02 14:58:12 [core.py:406]     self.model_executor = executor_class(vllm_config)
(AsyncvLLMServer pid=3620) ERROR 05-02 14:58:12 [core.py:406]   File "/home/ec2-user/vllm/vllm/executor/executor_base.py", line 52, in __init__
(AsyncvLLMServer pid=3620) ERROR 05-02 14:58:12 [core.py:406]     self._init_executor()
(AsyncvLLMServer pid=3620) ERROR 05-02 14:58:12 [core.py:406]   File "/home/ec2-user/verl/verl/workers/rollout/vllm_rollout/vllm_async_server.py", line 58, in _init_executor
(AsyncvLLMServer pid=3620) ERROR 05-02 14:58:12 [core.py:406]     assert len(actor_names) == vllm_dp_size * vllm_tp_size, f"instance_id: {self.vllm_config.instance_id} has {len(actor_names)} actors, but vllm_dp_size: {vllm_dp_size} * vllm_tp_size: {vllm_tp_size} = {vllm_dp_size * vllm_tp_size} is expected."
(AsyncvLLMServer pid=3620) ERROR 05-02 14:58:12 [core.py:406] AssertionError: instance_id: 39fff769-940c-4c88-bb08-0ad63fb5e321:5Y8l0k:1:0 has 0 actors, but vllm_dp_size: 1 * vllm_tp_size: 1 = 1 is expected.
(AsyncvLLMServer pid=3620) ERROR 05-02 14:58:12 [core.py:406] 
(raylet) A worker died or was killed while executing a task by an unexpected system error. To troubleshoot the problem, check the logs for the dead worker. RayTask ID: ffffffffffffffff4d25a17abd5c9be7ad69593a01000000 Worker ID: 5e1994c85b7e38c75d974f2886346b5662a2e33db3c8f160d894f1fb Node ID: d1cce7a158262a4f3567d99a4612737ffef92c57256a3bc1d555541b Worker IP address: xxx Worker port: 44215 Worker PID: 3620 Worker exit type: SYSTEM_ERROR Worker exit detail: Worker unexpectedly exits with a connection error code 2. End of file. There are some potential root causes. (1) The process is killed by SIGKILL by OOM killer due to high memory usage. (2) ray stop --force is called. (3) The worker is crashed unexpectedly due to SIGSEGV or other unexpected errors.
Traceback (most recent call last):
  File "/home/ec2-user/verl/tests/rollout/test_vllm_multi_turn.py", line 179, in <module>
    test_vllm_multi_turn()
  File "/home/ec2-user/verl/tests/rollout/test_vllm_multi_turn.py", line 100, in test_vllm_multi_turn
    async_rollout_manager = AsyncLLMServerManager(
  File "/home/ec2-user/verl/verl/workers/rollout/async_server.py", line 271, in __init__
    ray.get([worker.init_engine.remote() for worker in self.async_llm_servers])
  File "/usr/local/lib/python3.10/dist-packages/ray/_private/auto_init_hook.py", line 21, in auto_init_wrapper
    return fn(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/ray/_private/client_mode_hook.py", line 103, in wrapper
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/ray/_private/worker.py", line 2771, in get
    values, debugger_breakpoint = worker.get_objects(object_refs, timeout=timeout)
  File "/usr/local/lib/python3.10/dist-packages/ray/_private/worker.py", line 921, in get_objects
    raise value
ray.exceptions.ActorDiedError: The actor died unexpectedly before finishing this task.
        class_name: AsyncvLLMServer
        actor_id: 4d25a17abd5c9be7ad69593a01000000
        pid: 3620
        name: async_llm_worker_0
        namespace: 39fff769-940c-4c88-bb08-0ad63fb5e321
        ip: xxx
The actor is dead because its worker process has died. Worker exit type: SYSTEM_ERROR Worker exit detail: Worker unexpectedly exits with a connection error code 2. End of file. There are some potential root causes. (1) The process is killed by SIGKILL by OOM killer due to high memory usage. (2) ray stop --force is called. (3) The worker is crashed unexpectedly due to SIGSEGV or other unexpected errors.
root@ip-172-31-44-216:/home/ec2-user/verl# 
```
I correctly see the necessary Actors running in the Ray dashboard.

![Image](https://github.com/user-attachments/assets/428f3f89-8bf0-4f8d-b12a-d47db237c9cc)

## Appendix A
`verl/tests/rollout/test_vllm_multi_turn.py`
```python
# FIRST PART OF FILE UNCHANGED 

def test_vllm_multi_turn():
    config = OmegaConf.load("verl/trainer/config/ppo_trainer.yaml")
    model_path = "Qwen/Qwen2.5-0.5B-Instruct"
    model_name = "/".join(model_path.split("/")[-2:])
    config.actor_rollout_ref.model.path = model_path
    config.actor_rollout_ref.rollout.mode = "async"
    config.actor_rollout_ref.rollout.chat_scheduler = "examples.ppo_trainer.naive_chat_scheduler.NaiveChatCompletionScheduler"
    config.actor_rollout_ref.rollout.prompt_length = 2048
    config.actor_rollout_ref.rollout.response_length = 1024


    # test sleep/wake_up with fsdp offload
    config.actor_rollout_ref.actor.fsdp_config.param_offload = False
    config.actor_rollout_ref.actor.fsdp_config.optimizer_offload = False

    # add vm specific config
    config.ray_init.num_cpus = 8
    config.trainer.n_gpus_per_node = int(os.environ.get("NGPUS_PER_NODE", 1))
    config.actor_rollout_ref.rollout.tensor_model_parallel_size = int(os.environ.get("GEN_TP", 1))
    
    # =========================== 1. Create hybrid ActorRollout workers ===========================
    # make openai client happy
    os.environ["no_proxy"] = ""
    os.environ["http_proxy"] = ""
    os.environ["https_proxy"] = ""

    ray.init(
        logging_level=logging.DEBUG,
        runtime_env={
            "env_vars": {
                "TOKENIZERS_PARALLELISM": "true",
                "NCCL_DEBUG": "WARN",
                "VLLM_LOGGING_LEVEL": "WARN",
                "VLLM_USE_V1": "1",
            }
        }
    )
   
## REST OF FILE UNCHANGED
```

@ltdemey It's a little weird I can't reproduce the problem.
> AssertionError: instance_id: 39fff769-940c-4c88-bb08-0ad63fb5e321:5Y8l0k:1:0 has 0 actors, but vllm_dp_size: 1 * vllm_tp_size: 1 = 1 is expected.

This assertion says that AsyncvLLMServer can't find named ray actors starts with `5Y8l0k`(which should be launched by the actor_rollout_wg). Please add a debug message to print the named actor as below:
```diff
diff --git a/tests/rollout/test_vllm_multi_turn.py b/tests/rollout/test_vllm_multi_turn.py
index d37b8e8..5201416 100644
--- a/tests/rollout/test_vllm_multi_turn.py
+++ b/tests/rollout/test_vllm_multi_turn.py
@@ -42,6 +42,9 @@ def test_vllm_multi_turn():
     config.actor_rollout_ref.actor.fsdp_config.param_offload = True
     config.actor_rollout_ref.actor.fsdp_config.optimizer_offload = True
 
+    config.trainer.n_gpus_per_node = 1
+    config.actor_rollout_ref.rollout.tensor_model_parallel_size = 1
+
     # =========================== 1. Create hybrid ActorRollout workers ===========================
     # make openai client happy
     os.environ["no_proxy"] = ""
@@ -88,6 +91,8 @@ def test_vllm_multi_turn():
     actor_rollout_wg = all_wg["actor_rollout"]
     actor_rollout_wg.init_model()
 
+    actor_names = [actor_name for actor_name in ray.util.list_named_actors() if actor_name.startswith(f"{actor_rollout_wg.name_prefix}WorkerDict")]
+    print(f"[DEBUG]: actor_names: {actor_names}")
     # =========================== 2. Create AsyncLLMServerManager  ===========================
     async_rollout_manager = AsyncLLMServerManager(
         config=config.actor_rollout_ref,
```
Which should print message like
```text
WARNING:2025-05-03 10:49:06,656:Waiting for register center actor PP9ggc_register_center to be ready. Elapsed time: 0 seconds out of 300 seconds.
...
[DEBUG]: actor_names: ['PP9ggcWorkerDict_0:0']
...
```

Hm, it's indeed unexpected that you cannot recreate the error. I'm running on a simple `g6.2xlarge` ec2 instance.

In any case, the log accurately identifies that a ray actor is launched (`[DEBUG]: actor_names: ['misQrdWorkerDict_0:0']`). I also tried to add a `breakpoint()` at `verl/verl/workers/rollout/vllm_rollout/vllm_async_server.py:55` and notice that while the process of the worker is still running (see stack trace below), it is not listed under `ray.util.list_named_actors()`. 

```bash
Process 5480: ray::WorkerDict
Python v3.10.12 (/usr/bin/python3.10)

Thread 5480 (idle): "MainThread"
    main_loop (ray/_private/worker.py:935)
    <module> (ray/_private/workers/default_worker.py:297)
Thread 5636 (idle): "Thread-1 (_read_thread)"
    _recv_msg (torch/_inductor/compile_worker/subproc_pool.py:53)
    _read_thread (torch/_inductor/compile_worker/subproc_pool.py:161)
    run (threading.py:953)
    _bootstrap_inner (threading.py:1016)
    _bootstrap (threading.py:973)
```


`python tests/rollout/test_vllm_multi_turn.py`
```bash
2025-05-04 12:00:33,561 INFO worker.py:1832 -- Started a local Ray instance. View the dashboard at 127.0.0.1:8265 
[DEBUG] Main script Ray namespace: cc8e0b67-d5f3-4301-b358-282d67db13f6
2025-05-04 12:00:34,508 WARNING __init__.py:161 -- DeprecationWarning: `ray.state.available_resources_per_node` is a private attribute and access will be removed in a future Ray version.
colocated worker base class <class 'verl.single_controller.base.worker.Worker'>
bind role actor_rollout method execute_method to class <class 'verl.single_controller.ray.base.create_colocated_worker_cls.<locals>.WorkerDict'>
WARNING:2025-05-04 12:00:35,609:Waiting for register center actor rZXzt6_register_center to be ready. Elapsed time: 0 seconds out of 300 seconds.
(WorkerDict pid=2374) Model config after override: Qwen2Config {
(WorkerDict pid=2374)   "architectures": [
(WorkerDict pid=2374)     "Qwen2ForCausalLM"
(WorkerDict pid=2374)   ],
(WorkerDict pid=2374)   "attention_dropout": 0.0,
(WorkerDict pid=2374)   "eos_token_id": 151645,
(WorkerDict pid=2374)   "hidden_act": "silu",
(WorkerDict pid=2374)   "hidden_size": 896,
(WorkerDict pid=2374)   "initializer_range": 0.02,
(WorkerDict pid=2374)   "intermediate_size": 4864,
(WorkerDict pid=2374)   "max_position_embeddings": 32768,
(WorkerDict pid=2374)   "max_window_layers": 21,
(WorkerDict pid=2374)   "model_type": "qwen2",
(WorkerDict pid=2374)   "num_attention_heads": 14,
(WorkerDict pid=2374)   "num_hidden_layers": 24,
(WorkerDict pid=2374)   "num_key_value_heads": 2,
(WorkerDict pid=2374)   "pad_token_id": 151643,
(WorkerDict pid=2374)   "rms_norm_eps": 1e-06,
(WorkerDict pid=2374)   "rope_scaling": null,
(WorkerDict pid=2374)   "rope_theta": 1000000.0,
(WorkerDict pid=2374)   "sliding_window": 32768,
(WorkerDict pid=2374)   "tie_word_embeddings": true,
(WorkerDict pid=2374)   "torch_dtype": "bfloat16",
(WorkerDict pid=2374)   "transformers_version": "4.51.1",
(WorkerDict pid=2374)   "use_cache": true,
(WorkerDict pid=2374)   "use_sliding_window": false,
(WorkerDict pid=2374)   "vocab_size": 151936
(WorkerDict pid=2374) }
(WorkerDict pid=2374) 
(WorkerDict pid=2374) Flash Attention 2.0 only supports torch.float16 and torch.bfloat16 dtypes, but the current dype in Qwen2ForCausalLM is torch.float32. You should run training or inference using Automatic Mixed-Precision via the `with torch.autocast(device_type='torch_device'):` decorator, or load the model with the `torch_dtype` argument. Example: `model = AutoModel.from_pretrained("openai/whisper-tiny", attn_implementation="flash_attention_2", torch_dtype=torch.float16)`
(WorkerDict pid=2374) You are attempting to use Flash Attention 2.0 with a model not initialized on GPU. Make sure to move the model to GPU after initializing it on CPU with `model.to('cuda')`.
(WorkerDict pid=2374) [rank0]:[W504 12:00:43.714360945 ProcessGroupNCCL.cpp:4561] [PG ID 0 PG GUID 0 Rank 0]  using GPU 0 to perform barrier as devices used by this process are currently unknown. This can potentially cause a hang if this rank to GPU mapping is incorrect. Specify device_ids in barrier() to force use of a particular device, or call init_process_group() with a device_id.
(WorkerDict pid=2374) NCCL version 2.21.5+cuda12.4
(WorkerDict pid=2374) Qwen2ForCausalLM contains 494.03M parameters
(WorkerDict pid=2374) wrap_policy: functools.partial(<function _or_policy at 0x7f32a915e170>, policies=[functools.partial(<function transformer_auto_wrap_policy at 0x7f32a915e050>, transformer_layer_cls={<class 'transformers.models.qwen2.modeling_qwen2.Qwen2DecoderLayer'>})])
(WorkerDict pid=2374) /usr/local/lib/python3.10/dist-packages/torch/distributed/fsdp/_init_utils.py:444: UserWarning: FSDP is switching to use `NO_SHARD` instead of ShardingStrategy.FULL_SHARD since the world size is 1.
(WorkerDict pid=2374)   warnings.warn(
(WorkerDict pid=2374) Total steps: -1, num_warmup_steps: 0
(WorkerDict pid=2374) Actor use_remove_padding=False
[DEBUG]: actor_names: ['rZXzt6WorkerDict_0:0']
(WorkerDict pid=2374) /usr/local/lib/python3.10/dist-packages/torch/distributed/fsdp/fully_sharded_data_parallel.py:690: FutureWarning: FSDP.state_dict_type() and FSDP.set_state_dict_type() are being deprecated. Please use APIs, get_state_dict() and set_state_dict(), which can support different parallelisms, FSDP1, FSDP2, DDP. API doc: https://pytorch.org/docs/stable/distributed.checkpoint.html#torch.distributed.checkpoint.state_dict.get_state_dict .Tutorial: https://pytorch.org/tutorials/recipes/distributed_checkpoint_recipe.html .
(WorkerDict pid=2374)   warnings.warn(
INFO 05-04 12:00:47 [__init__.py:239] Automatically detected platform cuda.
(AsyncvLLMServer pid=2576) FastAPI startup
(AsyncvLLMServer pid=2576) override_generation_config: {'n': 1, 'logprobs': 0, 'max_tokens': 1024, 'temperature': 1.0, 'top_k': -1, 'top_p': 1, 'ignore_eos': False}
(AsyncvLLMServer pid=2576) WARNING 05-04 12:01:02 [arg_utils.py:1713] Detected VLLM_USE_V1=1 with Engine in background thread. Usage should be considered experimental. Please report any issues on Github.
(AsyncvLLMServer pid=2576) WARNING 05-04 12:01:02 [cuda.py:96] To see benefits of async output processing, enable CUDA graph. Since, enforce-eager is enabled, async output processor cannot be used
(AsyncvLLMServer pid=2576) WARNING 05-04 12:01:03 [core_client.py:368] SIGUSR1 handler not installed because we are not running in the main thread. In this case the forked engine process may not be killed when an exception is raised, and you need to handle the engine process shutdown manually.
(AsyncvLLMServer pid=2576) 2025-05-04 12:01:09,819      INFO worker.py:1832 -- Started a local Ray instance. View the dashboard at 127.0.0.1:8266 
(AsyncvLLMServer pid=2576) EngineCoreProc.run_engine_core
(AsyncvLLMServer pid=2576) EngineCoreProc.__init__
(AsyncvLLMServer pid=2576) EngineCore.__init__ - Start
(AsyncvLLMServer pid=2576) EngineCore.__init__ - Before model executor initialization
(AsyncvLLMServer pid=2576) namespace: cc8e0b67-d5f3-4301-b358-282d67db13f6
(AsyncvLLMServer pid=2576) actor_names: []
(AsyncvLLMServer pid=2576) [DEBUG]: all actors: []
(AsyncvLLMServer pid=2576) ERROR 05-04 12:01:10 [core.py:406] EngineCore hit an exception: Traceback (most recent call last):
(AsyncvLLMServer pid=2576) ERROR 05-04 12:01:10 [core.py:406]   File "/home/ec2-user/vllm/vllm/v1/engine/core.py", line 394, in run_engine_core
(AsyncvLLMServer pid=2576) ERROR 05-04 12:01:10 [core.py:406]     engine_core = EngineCoreProc(*args, **kwargs)
(AsyncvLLMServer pid=2576) ERROR 05-04 12:01:10 [core.py:406]   File "/home/ec2-user/vllm/vllm/v1/engine/core.py", line 335, in __init__
(AsyncvLLMServer pid=2576) ERROR 05-04 12:01:10 [core.py:406]     super().__init__(vllm_config, executor_class, log_stats)
(AsyncvLLMServer pid=2576) ERROR 05-04 12:01:10 [core.py:406]   File "/home/ec2-user/vllm/vllm/v1/engine/core.py", line 68, in __init__
(AsyncvLLMServer pid=2576) ERROR 05-04 12:01:10 [core.py:406]     self.model_executor = executor_class(vllm_config)
(AsyncvLLMServer pid=2576) ERROR 05-04 12:01:10 [core.py:406]   File "/home/ec2-user/vllm/vllm/executor/executor_base.py", line 52, in __init__
(AsyncvLLMServer pid=2576) ERROR 05-04 12:01:10 [core.py:406]     self._init_executor()
(AsyncvLLMServer pid=2576) ERROR 05-04 12:01:10 [core.py:406]   File "/home/ec2-user/verl/verl/workers/rollout/vllm_rollout/vllm_async_server.py", line 59, in _init_executor
(AsyncvLLMServer pid=2576) ERROR 05-04 12:01:10 [core.py:406]     assert len(actor_names) == vllm_dp_size * vllm_tp_size, f"instance_id: {self.vllm_config.instance_id} has {len(actor_names)} actors, but vllm_dp_size: {vllm_dp_size} * vllm_tp_size: {vllm_tp_size} = {vllm_dp_size * vllm_tp_size} is expected."
(AsyncvLLMServer pid=2576) ERROR 05-04 12:01:10 [core.py:406] AssertionError: instance_id: cc8e0b67-d5f3-4301-b358-282d67db13f6:rZXzt6:1:0 has 0 actors, but vllm_dp_size: 1 * vllm_tp_size: 1 = 1 is expected.
(AsyncvLLMServer pid=2576) ERROR 05-04 12:01:10 [core.py:406] 
(raylet) A worker died or was killed while executing a task by an unexpected system error. To troubleshoot the problem, check the logs for the dead worker. RayTask ID: ffffffffffffffff4002a1d9c4490b2eb5ecac7101000000 Worker ID: 482edc7ef16ece3f4e2b579296a684e36cdc0a098b8885b0cad01ffa Node ID: 67e823c9d9589801f9a45aa3731f818cd2e34862acf44d3ac73483f7 Worker IP address: 172.31.44.216 Worker port: 39343 Worker PID: 2576 Worker exit type: SYSTEM_ERROR Worker exit detail: Worker unexpectedly exits with a connection error code 2. End of file. There are some potential root causes. (1) The process is killed by SIGKILL by OOM killer due to high memory usage. (2) ray stop --force is called. (3) The worker is crashed unexpectedly due to SIGSEGV or other unexpected errors.
Traceback (most recent call last):
  File "/home/ec2-user/verl/tests/rollout/test_vllm_multi_turn.py", line 177, in <module>
    test_vllm_multi_turn()
  File "/home/ec2-user/verl/tests/rollout/test_vllm_multi_turn.py", line 98, in test_vllm_multi_turn
    async_rollout_manager = AsyncLLMServerManager(
  File "/home/ec2-user/verl/verl/workers/rollout/async_server.py", line 271, in __init__
    ray.get([worker.init_engine.remote() for worker in self.async_llm_servers])
  File "/usr/local/lib/python3.10/dist-packages/ray/_private/auto_init_hook.py", line 21, in auto_init_wrapper
    return fn(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/ray/_private/client_mode_hook.py", line 103, in wrapper
    return func(*args, **kwargs)
  File "/usr/local/lib/python3.10/dist-packages/ray/_private/worker.py", line 2771, in get
    values, debugger_breakpoint = worker.get_objects(object_refs, timeout=timeout)
  File "/usr/local/lib/python3.10/dist-packages/ray/_private/worker.py", line 921, in get_objects
    raise value
ray.exceptions.ActorDiedError: The actor died unexpectedly before finishing this task.
        class_name: AsyncvLLMServer
        actor_id: 4002a1d9c4490b2eb5ecac7101000000
        pid: 2576
        name: async_llm_worker_0
        namespace: cc8e0b67-d5f3-4301-b358-282d67db13f6
        ip: 172.31.44.216
The actor is dead because its worker process has died. Worker exit type: SYSTEM_ERROR Worker exit detail: Worker unexpectedly exits with a connection error code 2. End of file. There are some potential root causes. (1) The process is killed by SIGKILL by OOM killer due to high memory usage. (2) ray stop --force is called. (3) The worker is crashed unexpectedly due to SIGSEGV or other unexpected errors.
```

--- a/tests/rollout/async_rollout_utils.py
+++ b/tests/rollout/async_rollout_utils.py
@@ -12,7 +12,7 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 import os
-from typing import Any, Dict, Tuple
+from typing import Any, Dict
 
 import ray
 from omegaconf import DictConfig
@@ -24,7 +24,7 @@ from verl.workers.fsdp_workers import AsyncActorRolloutRefWorker
 from verl.workers.rollout.async_server import AsyncLLMServerManager
 
 
-def init_async_rollout_manager(config: DictConfig, scheduler_kwargs: Dict[str, Any] = None) -> Tuple[Dict[str, RayWorkerGroup], AsyncLLMServerManager]:
+def init_async_rollout_manager(config: DictConfig, scheduler_kwargs: Dict[str, Any] = None) -> AsyncLLMServerManager:
     # make openai client happy
     os.environ["no_proxy"] = ""
     os.environ["http_proxy"] = ""
@@ -51,13 +51,11 @@ def init_async_rollout_manager(config: DictConfig, scheduler_kwargs: Dict[str, A
     resource_pool_to_cls[resource_pool]["actor_rollout"] = actor_rollout_cls
 
     all_wg = {}
-    wg_dicts = []
     for resource_pool, class_dict in resource_pool_to_cls.items():
         worker_dict_cls = create_colocated_worker_cls(class_dict=class_dict)
         wg_dict = RayWorkerGroup(resource_pool=resource_pool, ray_cls_with_init=worker_dict_cls)
         spawn_wg = wg_dict.spawn(prefix_set=class_dict.keys())
         all_wg.update(spawn_wg)
-        wg_dicts.append(wg_dict)
     actor_rollout_wg = all_wg["actor_rollout"]
     actor_rollout_wg.init_model()
 
@@ -68,4 +66,4 @@ def init_async_rollout_manager(config: DictConfig, scheduler_kwargs: Dict[str, A
         scheduler_kwargs=scheduler_kwargs,
     )
 
-    return all_wg, async_rollout_manager
+    return async_rollout_manager

--- a/tests/rollout/test_vllm_multi_turn.py
+++ b/tests/rollout/test_vllm_multi_turn.py
@@ -55,7 +55,7 @@ def test_vllm_multi_turn(config):
 
     # =========================== 1. Init rollout manager ===========================
     model_name = "/".join(config.actor_rollout_ref.model.path.split("/")[-2:])
-    worker_groups, async_rollout_manager = init_async_rollout_manager(config)
+    async_rollout_manager = init_async_rollout_manager(config)
 
     # test sleep and wake_up
     async_rollout_manager.sleep()
@@ -145,7 +145,7 @@ async def test_vllm_streaming_response(config):
     )
 
     model_name = "/".join(config.actor_rollout_ref.model.path.split("/")[-2:])
-    worker_groups, async_rollout_manager = init_async_rollout_manager(config)
+    async_rollout_manager = init_async_rollout_manager(config)
     async_llm_server = async_rollout_manager.async_llm_servers[0]
 
     # non-streaming request

--- a/tests/rollout/test_vllm_tool_calling.py
+++ b/tests/rollout/test_vllm_tool_calling.py
@@ -264,7 +264,7 @@ def test_vllm_tool_calling():
     # Init sandbox and async rollout manager
     sandbox = Sandbox.options(num_cpus=1).remote()
     sandbox_address = ray.get(sandbox.get_server_address.remote())
-    worker_groups, async_rollout_manager = init_async_rollout_manager(config, scheduler_kwargs={"sandbox_address": sandbox_address, "system_prompt": system_prompt})
+    async_rollout_manager = init_async_rollout_manager(config, scheduler_kwargs={"sandbox_address": sandbox_address, "system_prompt": system_prompt})
 
     # Build dataset
     dataset = load_dataset("Maxwell-Jia/AIME_2024", split="train")

--- a/verl/single_controller/ray/base.py
+++ b/verl/single_controller/ray/base.py
@@ -189,6 +189,7 @@ class RayWorkerGroup(WorkerGroup):
         name_prefix: str = None,
         detached=False,
         worker_names=None,
+        worker_handles: List[ray.actor.ActorHandle] = None,
         ray_wait_register_center_timeout: int = 300,
         **kwargs,
     ) -> None:
@@ -206,7 +207,7 @@ class RayWorkerGroup(WorkerGroup):
             self._worker_names = worker_names
 
         if self._is_init_with_detached_workers:
-            self._init_with_detached_workers(worker_names=worker_names)
+            self._init_with_detached_workers(worker_names=worker_names, worker_handles=worker_handles)
         else:
             self._init_with_resource_pool(resource_pool=resource_pool, ray_cls_with_init=ray_cls_with_init, bin_pack=bin_pack, detached=detached)
 
@@ -220,8 +221,12 @@ class RayWorkerGroup(WorkerGroup):
         worker_state_dict = get_actor(worker._actor_id.hex())
         return worker_state_dict.get("state", "undefined") == "ALIVE" if worker_state_dict is not None else False
 
-    def _init_with_detached_workers(self, worker_names):
-        workers = [ray.get_actor(name=name) for name in worker_names]
+    def _init_with_detached_workers(self, worker_names, worker_handles):
+        # ray.get_actor holds a weak reference to the actor, which causes actors garbage collected unexpectedly
+        # if we only hold spawn RayWorkerGroup. By passing actor handle explicitly, spawn RayWorkerGroup have
+        # strong reference to these actors.
+        # https://github.com/ray-project/ray/pull/45699
+        workers = worker_handles if worker_handles else [ray.get_actor(name=name) for name in worker_names]
         self._workers = workers
         self._world_size = len(worker_names)
 
@@ -319,9 +324,10 @@ class RayWorkerGroup(WorkerGroup):
         cls,
         name_prefix,
         worker_names=None,
+        worker_handles=None,
         ray_cls_with_init=None,
     ):
-        worker_group = cls(resource_pool=None, ray_cls_with_init=ray_cls_with_init, name_prefix=name_prefix, worker_names=worker_names)
+        worker_group = cls(resource_pool=None, ray_cls_with_init=ray_cls_with_init, name_prefix=name_prefix, worker_names=worker_names, worker_handles=worker_handles)
         return worker_group
 
     def spawn(self, prefix_set):
@@ -349,6 +355,7 @@ class RayWorkerGroup(WorkerGroup):
             new_worker_group = self.from_detached(
                 name_prefix=self.name_prefix,
                 worker_names=self._worker_names,
+                worker_handles=self._workers,
                 ray_cls_with_init=self.ray_cls_with_init,
             )
 

--- a/verl/trainer/ppo/ray_trainer.py
+++ b/verl/trainer/ppo/ray_trainer.py
@@ -708,7 +708,6 @@ class RayPPOTrainer:
         # Instead, directly pass different resource pool to different worker groups.
         # See https://github.com/volcengine/verl/blob/master/examples/ray/tutorial.ipynb for more information.
         all_wg = {}
-        self.wg_dicts = []
         wg_kwargs = {}  # Setting up kwargs for RayWorkerGroup
         if OmegaConf.select(self.config.trainer, "ray_wait_register_center_timeout") is not None:
             wg_kwargs["ray_wait_register_center_timeout"] = self.config.trainer.ray_wait_register_center_timeout
@@ -718,8 +717,6 @@ class RayPPOTrainer:
             wg_dict = self.ray_worker_group_cls(resource_pool=resource_pool, ray_cls_with_init=worker_dict_cls, **wg_kwargs)
             spawn_wg = wg_dict.spawn(prefix_set=class_dict.keys())
             all_wg.update(spawn_wg)
-            # keep the referece of WorkerDict to support ray >= 2.31. Ref: https://github.com/ray-project/ray/pull/45699
-            self.wg_dicts.append(wg_dict)
 
         if self.use_critic:
             self.critic_wg = all_wg["critic"]

